cmd/updatetree/main_test.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/trillian"
+	"github.com/google/trillian/testonly"
+	"github.com/google/trillian/util/flagsaver"
+)
+
+type testCase struct {
+	desc       string
+	setFlags   func()
+	updateErr  error
+	wantRPC    bool
+	updateTree *trillian.Tree
+	wantErr    bool
+	wantState  trillian.TreeState
+}
+
+func TestFreezeTree(t *testing.T) {
+	runTest(t, []*testCase{
+		{
+			// We don't set the treeID in runTest so this should fail.
+			desc:    "missingTreeID",
+			wantErr: true,
+		},
+		{
+			desc: "mandatoryOptsNotSet",
+			// Undo the flags set by runTest, so that mandatory options are no longer set.
+			setFlags: resetFlags,
+			wantErr:  true,
+		},
+		{
+			desc: "validUpdateFrozen",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC: true,
+			updateTree: &trillian.Tree{
+				TreeId:    12345,
+				TreeState: trillian.TreeState_FROZEN,
+			},
+			wantState: trillian.TreeState_FROZEN,
+		},
+		{
+			desc: "updateInvalidState",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "ITSCOLDOUTSIDE"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "unknownTree",
+			setFlags: func() {
+				*treeID = 123456
+				*treeState = "FROZEN"
+			},
+			wantErr:   true,
+			wantRPC:   true,
+			updateErr: errors.New("unknown tree id"),
+		},
+		{
+			desc: "emptyAddr",
+			setFlags: func() {
+				*adminServerAddr = ""
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantErr: true,
+		},
+		{
+			desc: "updateErr",
+			setFlags: func() {
+				*treeID = 12345
+				*treeState = "FROZEN"
+			},
+			wantRPC:   true,
+			updateErr: errors.New("update tree failed"),
+			wantErr:   true,
+		},
+	})
+}
+
+// runTest executes the updateTree command against a fake TrillianAdminServer
```

Comment:
Not sure `runTest` is adding much compared to just having the loop inline in the `Test..` function -- are other things going to use it?

Reply:
Again it does slightly more work in createtree and was retained to make them look similar.